File: python/02_inmutabilidad/inmutabilidad.py
"""
INMUTABILIDAD
=============
En programación funcional los datos no se modifican:
en lugar de mutar una estructura, se crea una nueva versión.

Python no impone inmutabilidad, pero la podemos modelar con:
  - Tuplas en vez de listas
  - namedtuple / dataclass(frozen=True)
  - Patrones de copia-y-actualización
"""

from dataclasses import dataclass
from typing import Tuple

# -----------------------------------------------
# MUTABLE (estilo imperativo) ← evitar
# -----------------------------------------------
def agregar_impuesto_mutable(precios: list, tasa: float) -> list:
    for i in range(len(precios)):
        precios[i] *= (1 + tasa)   # MUTA la lista original!
    return precios


# -----------------------------------------------
# INMUTABLE (estilo funcional) ← preferir
# -----------------------------------------------
def agregar_impuesto(precios: Tuple[float, ...], tasa: float) -> Tuple[float, ...]:
    return tuple(p * (1 + tasa) for p in precios)


# -----------------------------------------------
# dataclass inmutable con frozen=True
# -----------------------------------------------
@dataclass(frozen=True)
class Punto:
    x: float
    y: float

    def mover(self, dx: float, dy: float) -> "Punto":
        # No muta self, retorna un NUEVO Punto
        return Punto(self.x + dx, self.y + dy)

    def __repr__(self):
        return f"Punto({self.x}, {self.y})"


# -----------------------------------------------
# DEMO
# -----------------------------------------------
if __name__ == "__main__":
    print("=== Mutable vs Inmutable ===")
    precios_originales = [100.0, 200.0, 50.0]
    print(f"Antes (mutable): {precios_originales}")
    agregar_impuesto_mutable(precios_originales, 0.19)
    print(f"Después (mutable): {precios_originales}")  # ¡lista modificada!

    print()
    precios_tupla = (100.0, 200.0, 50.0)
    nuevos_precios = agregar_impuesto(precios_tupla, 0.19)
    print(f"Original (inmutable): {precios_tupla}")    # sin cambios
    print(f"Con impuesto:         {nuevos_precios}")

    print("\n=== Punto inmutable ===")
    p1 = Punto(0, 0)
    p2 = p1.mover(3, 4)
    print(f"p1 = {p1}")   # sigue siendo (0, 0)
    print(f"p2 = {p2}")   # nuevo punto (3, 4)

    # Intentar mutar dispara error:
    try:
        p1.x = 10
    except Exception as e:
        print(f"Error al mutar: {e}")

File: java/src/main/java/com/paradigmas/fp/inmutabilidad/Inmutabilidad.java
package com.paradigmas.fp.inmutabilidad;

import java.util.List;
import java.util.stream.Collectors;

/**
 * INMUTABILIDAD
 * =============
 * En programación funcional los datos no se modifican:
 * se crea una nueva versión en lugar de mutar la existente.
 *
 * Java 26 ofrece:
 *   - Records  → clases de datos inmutables por diseño
 *   - List.of() / List.copyOf() → listas no modificables
 *   - Colecciones inmutables con Collections.unmodifiableXxx()
 */
public class Inmutabilidad {

    // ── RECORD: inmutable por definición ─────────────────────────────
    /**
     * Record Punto: campos finales, constructor canónico, equals/hashCode/toString
     * generados automáticamente. No tiene setters → inmutable.
     */
    record Punto(double x, double y) {
        /** En lugar de mutar, retorna un NUEVO Punto desplazado. */
        Punto mover(double dx, double dy) {
            return new Punto(x + dx, y + dy);
        }

        /** Distancia al origen (función pura sobre el record). */
        double distanciaAlOrigen() {
            return Math.sqrt(x * x + y * y);
        }
    }

    // ── RECORD con validación ─────────────────────────────────────────
    record Precio(double valor, String moneda) {
        // Compact constructor con validación
        Precio {
            if (valor < 0) throw new IllegalArgumentException("El precio no puede ser negativo");
            moneda = moneda.toUpperCase();
        }

        Precio aplicarDescuento(double porcentaje) {
            return new Precio(valor * (1 - porcentaje / 100), moneda);
        }

        Precio convertir(String nuevaMoneda, double tasa) {
            return new Precio(valor * tasa, nuevaMoneda);
        }
    }

    // ── MUTABLE (estilo imperativo) ← evitar ─────────────────────────
    static void agregarImpuestoMutable(double[] precios, double tasa) {
        for (int i = 0; i < precios.length; i++) {
            precios[i] *= (1 + tasa);   // MUTA el array original!
        }
    }

    // ── INMUTABLE (estilo funcional) ← preferir ──────────────────────
    static List<Double> agregarImpuesto(List<Double> precios, double tasa) {
        return precios.stream()
                      .map(p -> p * (1 + tasa))
                      .toList();   // nueva lista, original intacta
    }

    // ── DEMO ─────────────────────────────────────────────────────────
    public static void main(String[] args) {
        System.out.println("=== Record Punto (inmutable) ===");
        var p1 = new Punto(0, 0);
        var p2 = p1.mover(3, 4);
        System.out.println("p1 = " + p1);   // Punto[x=0.0, y=0.0]
        System.out.println("p2 = " + p2);   // Punto[x=3.0, y=4.0]
        System.out.println("distancia p2: " + p2.distanciaAlOrigen());

        System.out.println("\n=== Record Precio con validación ===");
        var precio = new Precio(1000, "usd");
        var conDescuento = precio.aplicarDescuento(20);
        var enEuros = conDescuento.convertir("EUR", 0.92);
        System.out.println("Original:     " + precio);
        System.out.println("Con 20% dto:  " + conDescuento);
        System.out.println("En EUR:       " + enEuros);

        System.out.println("\n=== Lista mutable vs inmutable ===");
        double[] preciosMutable = {100.0, 200.0, 50.0};
        System.out.println("Antes: " + java.util.Arrays.toString(preciosMutable));
        agregarImpuestoMutable(preciosMutable, 0.19);
        System.out.println("Después (mutable): " + java.util.Arrays.toString(preciosMutable)); // cambiada

        List<Double> preciosOrig = List.of(100.0, 200.0, 50.0);
        List<Double> preciosNuevos = agregarImpuesto(preciosOrig, 0.19);
        System.out.println("Original (inmutable): " + preciosOrig);   // sin cambios
        System.out.println("Con impuesto:         " + preciosNuevos);
    }
}
